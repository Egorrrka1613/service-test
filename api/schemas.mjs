import joi from 'joi';

export const patterns = {
    evmAddress: /^0x([a-zA-Z0-9]{40})$/,
    numeric: /^\d+(\.\d+)?$/,
};

export const types = {
    transaction: joi.object({
        from: joi.string().pattern(patterns.evmAddress).required(),
        to: joi.string().pattern(patterns.evmAddress).required(),
        data: joi.string().required(),
        value: joi.string().pattern(patterns.numeric).required(),
        nonce: joi.number(),
        gasPrice: joi.string().pattern(patterns.numeric).required(),
        gas: joi.string().pattern(patterns.numeric).required(),
        chainId: joi.number().required(),
    }),
    amount: joi.object({
        amount: joi.string().pattern(patterns.numeric).required(),
        currency: joi.string().required(),
    }),
}

