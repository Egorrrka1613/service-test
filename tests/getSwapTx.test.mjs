import { getSwapTx } from "../api/index.mjs";
import { types } from '../api/schemas.mjs';
import { validateArgs } from '../api/helpers/index.mjs';
import assertValidate from './helpers/assertValidate.mjs';
import { evm } from '@zomet/sdk';
import { jest } from '@jest/globals'
import nock from 'nock';

let inputData;
beforeEach(() => {
    nock.cleanAll();
    inputData = {
        net: 'eth',
        fromTokenAddress: '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee',
        toTokenAddress: '0xdac17f958d2ee523a2206206994597c13d831ec7',
        ownerAddress: '0xaac594a9b83f892f1a4b5b8804339fc6ddb797ee',
        amount: '0.0001',
        slippage: '0.5'
    }
});

afterAll(() => {
    nock.restore();
});


describe('getSwapTx method tests', () => {
    it('Case #1: get tx swap from Ether to USDT in ETH', async () => {
        // given
        const expectResponseByHttpMock = {
            fromToken: {
                symbol: 'ETH',
                name: 'Ether',
                decimals: 18,
                address: inputData.fromTokenAddress,
                logoURI: 'https://tokens.1inch.io/0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee.png',
                tags: [Array]
            },
            toToken: {
                symbol: 'USDT',
                name: 'Tether USD',
                address: inputData.toTokenAddress,
                decimals: 6,
                logoURI: 'https://tokens.1inch.io/0xdac17f958d2ee523a2206206994597c13d831ec7.png',
                tags: [Array]
            },
            toTokenAmount: '181177',
            fromTokenAmount: '100000000000000',
            protocols: [[Array]],
            tx: {
                from: inputData.ownerAddress,
                to: '0x1111111254eeb25477b68fb85ed929f73a960582',
                data: '0x0502b1c5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005af3107a4000000000000000000000000000000000000000000000000000000000000002c02f0000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000100000000000000003b6d03400d4a11d5eeaac28ec3f61d100daf4d40471f185261c0064f',
                value: '100000000000000',
                gas: 145146,
                gasPrice: '18069299540'
            }
        };

        const txMock = {
            from: inputData.ownerAddress,
            to: "0x1111111254eeb25477b68fb85ed929f73a960582",
            data: "0x0502b1c5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005af3107a4000000000000000000000000000000000000000000000000000000000000002c02f0000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000100000000000000003b6d03400d4a11d5eeaac28ec3f61d100daf4d40471f1852",
            value: "100000000000000000000000000000000",
            gasPrice: "18069299540",
            gas: "145146",
            chainId: 1,
        }

        const inchRouterContractMock = { buildTransaction: jest.fn().mockReturnValue(txMock) };
        jest.spyOn(evm, 'Contract').mockReturnValue(inchRouterContractMock);

        const httpMock = nock(`${process.env.INCH_API_URL}`)
            .get('/v5.0/1/swap')
            .query(true)
            .reply(200, expectResponseByHttpMock);

        // when
        const response = await getSwapTx(inputData);

        // then
        expect(httpMock.isDone()).toBe(true);
        validateArgs(response, types.transaction);
        expect(response).toMatchSnapshot();
    });

    it(`Case #2: Assert error if use invalid method signature`,
        async () => assertValidate(getSwapTx, inputData));

    it('Case #3: Assert error if use invalid network name', async () => {
        inputData.net = 'abc';
        await expect(getSwapTx(inputData)).rejects.toMatchSnapshot();
    });

    it('Case #4: Assert service error if send wrong contract in toTokenAddress', async () => {
        const expectResponseByMock = `${inputData.toTokenAddress} is wrong address`;

        const mock = nock(`${process.env.INCH_API_URL}`)
            .get('/v5.0/1/swap')
            .query(true)
            .reply(400, expectResponseByMock);

        await expect(getSwapTx(inputData)).rejects.toEqual(expectResponseByMock);
        expect(mock.isDone()).toBe(true);
    });

    it(`Case #5: Assert error if use invalid param "fromTokenAddress"`, async () => {
        inputData.fromTokenAddress = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
        await expect(getSwapTx(inputData)).rejects.toMatchSnapshot();
    });
});
