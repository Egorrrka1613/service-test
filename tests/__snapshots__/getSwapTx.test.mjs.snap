// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSwapTx method tests Case #1: get tx swap from Ether to USDT in ETH 1`] = `
{
  "chainId": 1,
  "data": "0x0502b1c5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005af3107a4000000000000000000000000000000000000000000000000000000000000002c02f0000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000100000000000000003b6d03400d4a11d5eeaac28ec3f61d100daf4d40471f185261c0064f",
  "from": "0xaac594a9b83f892f1a4b5b8804339fc6ddb797ee",
  "gas": "145146",
  "gasPrice": "18069299540",
  "to": "0x1111111254eeb25477b68fb85ed929f73a960582",
  "value": "100000000000000",
}
`;

exports[`getSwapTx method tests Case #2: Assert error if use invalid method signature 1`] = `""net" is required"`;

exports[`getSwapTx method tests Case #2: Assert error if use invalid method signature 2`] = `""fromTokenAddress" is required"`;

exports[`getSwapTx method tests Case #2: Assert error if use invalid method signature 3`] = `""toTokenAddress" is required"`;

exports[`getSwapTx method tests Case #2: Assert error if use invalid method signature 4`] = `""ownerAddress" is required"`;

exports[`getSwapTx method tests Case #2: Assert error if use invalid method signature 5`] = `""amount" is required"`;

exports[`getSwapTx method tests Case #2: Assert error if use invalid method signature 6`] = `""slippage" is required"`;

exports[`getSwapTx method tests Case #3: Assert error if use invalid network name 1`] = `""net" must be one of [eth, bsc, polygon, arbitrum, avalanche, optimism]"`;

exports[`getSwapTx method tests Case #5: Assert error if use invalid param "fromTokenAddress" 1`] = `""fromTokenAddress" with value "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee" fails to match the required pattern: /^0x([a-zA-Z0-9]{40})$/"`;
